Tests/UpstreamIntegrationTests/UpstreamIntegrationTests.swift

```diff
@@ -194,5 +308,133 @@ class UpstreamIntegrationTests: XCTestCase {
     func registerEdgeLocationHintListener(_ listener: @escaping EventListener) {
         MobileCore.registerEventListener(type: FunctionalTestConst.EventType.EDGE, source: "locationHint:result", listener: listener)
     }
+    
+    // MARK: AnyCodable helpers
+    /// Performs testing assertions between two `[AnyCodable]` instances.
+    func assertEqual(lhs: [String: AnyCodable]?, rhs: [String: AnyCodable]?, keyPath: [String]) {
+        if lhs == nil, rhs == nil {
+            return
+        }
+        guard let lhs = lhs, let rhs = rhs else {
+            XCTFail(#"""
+                \#(lhs == nil ? "lhs is nil" : "rhs is nil") and \#(lhs == nil ? "rhs" : "lhs") is non-nil
+
+                lhs: \#(String(describing: lhs))
+                
+                rhs: \#(String(describing: rhs))
+                
+                key path: \#(keyPath)
+            """#)
+            return
+        }
+        if lhs.count != rhs.count {
+            XCTFail(#"""
+                lhs and rhs counts do not match.
+                lhs count: \#(lhs.count)
+                rhs count: \#(rhs.count)
+                
+                lhs: \#(lhs)
+                
+                rhs: \#(rhs)
+                
+                key path: \#(keyPath)
```

Comment:
I have a small improvement request to prettify the keyPath instead of printing the array directly. Key path examples:

| Current | Prettified |
|---|-----|
| ["integerCompare"] | integerCompare |
| ["arrayOfObjectsFail", "[1]", "object2"] | arrayOfObjectsFail[1].object2 |
| ["dictionaryNestedCompare", "nested1", "nested2", "nested3", "nested4", "nested1"] | dictionaryNestedCompare.nested1.nested2.nested3.nested4.nested1 |

Otherwise the failure messages look great 😍

Reply:
Quotes look good to me and this is also in line with how you access keys with special formatting in JS. And I like your suggestion to display them only when the key is not composed of alphanumerics and non-reserved characters (may include space, dot, *, brackets - or others you may use in your formatting). 
Few examples can be:

`key1 -> key2` toString: `key1.key2` (no quotes)
`key1.key2 -> key3`  toString: `"key1.key2".key3`
`key 1 -> key 2` toString: `"key 1"."key 2"`
`key***1 -> key2` -> [0] toString: `"key***1".key2[0]`
`key[example] -> key2` toString: `"key[example]".key2`